parlai/mturk/tasks/turn_annotations/bot_agent.py

```diff
@@ -0,0 +1,175 @@
+        else:
+            act_out = self.model_agent.act()
+
+        annotations_html = TurkLikeAgent.construct_annotations_html(self.turn_idx)
+
+        if 'dict_lower' in self.opt and not self.opt['dict_lower']:
+            # model is cased so we don't want to normalize the reply like below
+            final_message_text = act_out['text']
+        else:
+            normalized_act_text = normalize_reply(act_out['text'])
+            final_message_text = normalized_act_text + annotations_html
+
+        if self.turn_idx >= self.num_turns * 2:
+            radio_css_style = 'margin-left:5px;margin-right:15px;'
+            radio_buttons_html = ''
+            for i in range(1, 6):
+                radio_buttons_html += f"""<input type="radio" id="radio_rating_{i}" name="radio_final_rating_group" value="{i}" /><span style={radio_css_style}>{i}</span>"""
+            exceeds_min_turns = f"""<br><br><div>{self.num_turns} chat turns finished! Please rate your partner on a scale of 1-5, how much would you enjoy talking to this partner over the course of a long conversation? (1: not at all, 5: a lot)</div>
+            {radio_buttons_html}
+            <br>Then, please click the "Done" button to end the chat."""
+            final_message_text += exceeds_min_turns
+            act_out = Compatibility.backward_compatible_force_set(
+                act_out, 'exceed_min_turns', True
+            )
+
+        act_out = Compatibility.backward_compatible_force_set(
+            act_out, 'text', final_message_text
+        )
+        assert ('episode_done' not in act_out) or (not act_out['episode_done'])
+        self.turn_idx += 1
+        return {**act_out, 'episode_done': False, 'checked_radio_name_id': ''}
+
+    def observe(self, observation):
+        """
+        Need to protect the observe also with a semaphore for composed models where an
+        act() may be called within an observe()
+        """
+        print(
+            f'{self.__class__.__name__}: In observe() before semaphore, self.turn_idx is {self.turn_idx} and observation is {observation}'
+        )
+        new_ob = copy.deepcopy(observation)
+        if self.semaphore:
+            with self.semaphore:
+                self.model_agent.observe(new_ob)
+        else:
+            self.model_agent.observe(new_ob)
+        print(
+            f'{self.__class__.__name__}: In observe() AFTER semaphore, self.turn_idx: {self.turn_idx}, observation["text"]: {new_ob["text"]}'
+        )
+
+        self.turn_idx += 1
+
+    def shutdown(self):
+        pass
+
+    def reset(self):
+        self.model_agent.reset()
+
+    @staticmethod
+    def get_bot_agents(opt: dict, active_models: list, datapath: str, no_cuda=False):
+        model_overrides = {
+            'datatype': 'valid',  # So we don't have to load the optimizer
+            'encode_candidate_vecs': True,  # For pulling from fixed list cands
+            'interactive_mode': True,
+            'model_parallel': True,
```

Comment:
we might want to flag this or make it a constant at the top or something

Reply:
Would there be anything wrong with just changing this to opt['model_parallel'] ?

Reply:
Answered my own question. Adding a new param.